### PR TITLE
add marathon scheduler for heron

### PR DESCRIPTION
**Pull request by nlu90**
This PR adds Marathon Scheduler for Heron.

### CODE REVIEW

**Comment by maosongfu** [comment]
Can u add the unit tests for the Marathon Scheduler?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+      conn.disconnect();
+    }
+
+    return true;
+  }
+
+  public boolean restartApp(int appId) {
+    if (appId == -1) {
+      // TODO (nlu): implement restart all
+      throw new RuntimeException("Restart all containers not supported yet");
```
Comment (maosongfu) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: Can we simply kill the whole group and resubmit the whole group in this case?

File: heron/spi/src/java/com/twitter/heron/spi/utils/SchedulerUtils.java [hunk 2]
```diff
@@ -438,4 +438,20 @@ public static boolean curlAndExtractPackage(
 
     return true;
   }
+
+  /**
+   * This method finds the container with highest resource requirement and returns the resource.
+   * Currently only RAM is used for max identification.
+   */
+  public static PackingPlan.Resource getMaxRequiredResource(PackingPlan packingPlan) {
```
Comment (maosongfu) [inline heron/spi/src/java/com/twitter/heron/spi/utils/SchedulerUtils.java]: So only RAM is taken into account? Should we rename it to reduce confusion?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  // submit a topology as a group, containers as apps in the group
+  public boolean submitTopology(String appConf) {
+    if (this.isVerbose) {
+      LOG.log(Level.INFO, "Topology conf is: " + appConf);
+    }
+
+    // Setup Connection
+    String schedulerURI = String.format("%s/v2/groups", this.marathonURI);
```
Comment (maosongfu) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: Why not use or refactor the utils method already exists?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraScheduler.java [hunk 4]
```diff
@@ -131,7 +132,7 @@ protected String getHeronAuroraPath() {
     TopologyAPI.Topology topology = Runtime.topology(runtime);
 
     // Align the ram to the maximal one
-    PackingPlan.Resource containerResource = getMaxRequiredResource(packing);
+    PackingPlan.Resource containerResource = SchedulerUtils.getMaxRequiredResource(packing);
```
Comment (maosongfu) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraScheduler.java]: It is better to put this change into a separate pull request.

**Comment by maosongfu** [comment]
Can we also add docs for this Marathon Scheduler?

**Comment by kramasamy** [comment]
@maosongfu - can the documentation be added in a separate PR? @nlu90 - please create another issue to track this.

**Comment by nlu90** [comment]
@kramasamy Created the documentation issue #1173

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 1 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: I left this unimplemented because I could not find a marathon api to restart a complete group and wondering if we can get some instructions from the Marathon people.

Yes, we can kill the group and then resubmit it. The only thing is we need to fetch the group conf first, but luckily marathon provide such api.

File: heron/spi/src/java/com/twitter/heron/spi/utils/SchedulerUtils.java (hunk 2 above)
Comment (nlu90) [inline heron/spi/src/java/com/twitter/heron/spi/utils/SchedulerUtils.java]: This method is moved from AuroraScheduler, maybe we should put this refactor into a separate PR.

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 3 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: Those utils methods need to be refactored for usage. And to reduce the development time, I choose to write my own method for the initial commit. 

I'll refactor and use them in the following update.

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 1 above)
Comment (mhausenblas) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: Yes, see https://mesosphere.github.io/marathon/docs/generated/api.html#v2_groups_delete

Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: @mhausenblas, Is there any marathon api to get the application group JSON file when the group is sbumitted?

I tried `GET /v2/groups/{groupId}`, but the returned result is not the original JSON file, it modifies the "id" fields, adds some extra information that not accepted during submission.

**Comment by kramasamy** [comment]
@nlu90 - you might have to resolve conflicts.

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 3 above)
Comment (kramasamy) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: can you open a PR for this?

File: heron/spi/src/java/com/twitter/heron/spi/utils/NetworkUtils.java [hunk 5]
```diff
@@ -142,27 +147,32 @@ public static boolean sendHttpPostRequest(HttpURLConnection connection,
...
-        LOG.log(Level.SEVERE, "Failed to close OutputStream: ", e);
+        LOG.log(Level.SEVERE, "Failed to send request: ", e);
         return false;
+      } finally {
+        try {
+          if (os != null) {
+            os.close();
+          }
+        } catch (IOException e) {
+          LOG.log(Level.SEVERE, "Failed to close OutputStream: ", e);
```
Comment (osgigeek) [inline heron/spi/src/java/com/twitter/heron/spi/utils/NetworkUtils.java]: What is the expectation here for logging this message? It becomes very difficult to know if there was a real issue with a stream which cannot be closed or failed on housekeeping tasks to cleanup resources. Preferably we should just use something like `IOUtils.closeQuietly(os)`

File: heron/spi/src/java/com/twitter/heron/spi/utils/NetworkUtils.java [hunk 6]
```diff
@@ -229,9 +250,25 @@ public static boolean sendHttpGetRequest(HttpURLConnection connection) {
...
   }
 
-  public static HttpURLConnection getConnection(String endpoint) throws IOException {
-    URL url = new URL(endpoint);
-    return (HttpURLConnection) url.openConnection();
+  public static HttpURLConnection getHttpConnection(String endpoint) {
+    // construct the http url connection
+    try {
+      URL url = new URL(endpoint);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
```
Comment (osgigeek) [inline heron/spi/src/java/com/twitter/heron/spi/utils/NetworkUtils.java]: We need some defensive coding here before assuming that the connection is a HTTPURLConnection

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 3 above)
Comment (osgigeek) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: @nlu90 if you refactor the methods it would be good to keep the connection creation (preferably using a connection pool) within the class which is doing all the additional work of setting Request-type, etc. That might also mean not having the flexibility to use static methods.

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonScheduler.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    String completeSchedulerProcessClassPath = new StringBuilder()
+        .append(Context.schedulerSandboxClassPath(config)).append(":")
+        .append(Context.packingSandboxClassPath(config)).append(":")
+        .append(Context.stateManagerSandboxClassPath(config))
+        .toString();
+    commands.add(completeSchedulerProcessClassPath);
+    commands.add(MarathonConstants.SCHEDULER_PORT);
+
+    return Joiner.on(" ").join(commands);
+  }
```
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonScheduler.java]: This method duplicate most of the code from `SchedulerUtils.executorCommandArgs`. Can we reuse the original method? It may need some refactoring though.

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    if (!NetworkUtils.sendHttpPostRequest(conn, NetworkUtils.JSON_TYPE, appConf.getBytes())) {
+      LOG.log(Level.SEVERE, "Failed to ");
+      conn.disconnect();
+      return false;
+    }
+
+    // Check response
+    boolean success = NetworkUtils.checkHttpResponseCode(conn, HttpURLConnection.HTTP_CREATED);
+    // Disconnect to release resources
+    conn.disconnect();
```
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: Should disconnect be put in a finally block?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    HttpURLConnection conn = NetworkUtils.getHttpConnection(topologyURI);
+    if (conn == null) {
+      LOG.log(Level.SEVERE, "Failed to find marathon scheduler");
+      return false;
+    }
+
+    // Send kill topology request
+    if (!NetworkUtils.sendHttpDeleteRequest(conn)) {
+      LOG.log(Level.SEVERE, "Failed to set delete request");
+      return false;
```
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: connection.disconnect?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    // Check response
+    boolean success = NetworkUtils.checkHttpResponseCode(conn, HttpURLConnection.HTTP_OK);
+    // Disconnect to release resources
+    conn.disconnect();
```
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: I think a finally block is needed for disconnect?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+package com.twitter.heron.scheduler.marathon;
+
+import java.net.HttpURLConnection;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import com.twitter.heron.spi.utils.NetworkUtils;
+
+public class MarathonController {
```
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: I am not sure I understand why this class called a `Controller`. It seems more like a "util" class to me. Moreover the http calls and related code could actually be moved to the `MarathonScheduler`. Also there is sole level of code duplication between the methods.

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    }
+  }
+
+  public boolean restartApp(int appId) {
+    if (appId == -1) {
+      // TODO (nlu): implement restart all
+      String message = "Restarting the whole topology is not supported yet. "
+          + "Please kill and resubmit the topology.";
+      LOG.log(Level.SEVERE, message);
+      return true;
```
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: Should the return value be false to let the user know that `restartAll` failed?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonScheduler.java (hunk 7 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonScheduler.java]: Yeah, I can refactor the method to be reused

File: heron/spi/src/java/com/twitter/heron/spi/utils/NetworkUtils.java (hunk 5 above)
Comment (nlu90) [inline heron/spi/src/java/com/twitter/heron/spi/utils/NetworkUtils.java]: Sounds good, we can make the close() be quiet.

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+  ) {
+    this.marathonURI = marathonURI;
+    this.topologyName = topologyName;
+    this.isVerbose = isVerbose;
+  }
+
+  public boolean killTopology() {
+    // Setup Connection
+    String topologyURI = String.format("%s/v2/groups/%s", this.marathonURI, this.topologyName);
+    HttpURLConnection conn = NetworkUtils.getHttpConnection(topologyURI);
```
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: I am not an expert. So wanted to clarify one doubt. Can Marathon service be SSL enabled? If so do we need to provide support for https?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 10 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: Do you mean, instead of do disconnect after each failure case, just `try {} finally {conn.disconnect()}`?

One concern is, all the methods called here don't throw any exception. In this case, should we still use `try` ?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 3 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: @kramasamy @osgigeek the refactor is done in this PR. For each request type, we have different util methods taking care of all the additional works

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 10 above)
Comment (ashvina) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: I was thinking more like the first one; one try finally for each method. 

You are correct, handling IOException captures most of the error cases. A try-finally is a generally accepted pattern whenever resources are concerned. I would suggest try-finally just to be sure no return paths are executed without closing the connection.

Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: sounds good, I'll update to use the try-finally pattern

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 12 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: The problem with returning false here is the RuntimeManagerMain would throw exceptions claiming the operation failed. If throwing exception instead of show some log message is acceptable, we can return true

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 13 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: I'm also not very sure. @mhausenblas could you explain the SSL a little bit?

File: heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java (hunk 11 above)
Comment (nlu90) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java]: The `Controller` is the one do the actual action. The reason I moved the logic out is to make the code more modularized and understandable.

**Comment by ashvina** [comment]
LGTM 👍 

@nlu90 do you think it is worth creating an issue for SSL?

**Comment by kramasamy** [comment]
👍 As @ashvina mentioned, please create the SSL issue - which we can resolve later.
